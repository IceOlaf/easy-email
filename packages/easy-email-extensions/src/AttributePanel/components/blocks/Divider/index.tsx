import React from 'react';
import { Padding } from '@extensions/AttributePanel/components/attributes/Padding';
import { ContainerBackgroundColor } from '@extensions/AttributePanel/components/attributes/ContainerBackgroundColor';
import { BorderWidth } from '@extensions/AttributePanel/components/attributes/BorderWidth';
import { BorderStyle } from '@extensions/AttributePanel/components/attributes/BorderStyle';
import { BorderColor } from '@extensions/AttributePanel/components/attributes/BorderColor';
import { Width } from '@extensions/AttributePanel/components/attributes/Width';
import { Align } from '@extensions/AttributePanel/components/attributes/Align';

import { AttributesPanelWrapper } from '@extensions/AttributePanel/components/attributes/AttributesPanelWrapper';
import { Collapse } from 'antd';
import { Stack } from 'easy-email-editor';

export function Divider() {
  return (

    <AttributesPanelWrapper>
      <Collapse defaultActiveKey={['-1', '0', '1', '2', '3']}>

        <Collapse.Panel key="1" header="Dimension">
          <Width inline />
          <Align />
          <Padding />
        </Collapse.Panel>

        <Collapse.Panel key="2" header="Border">
          <Stack wrap={false} spacing="tight">
            <div style={{ width: 50 }}><BorderWidth /></div>
            <div style={{ width: 100 }}><BorderStyle /></div>
            <div style={{ width: 100 }}><BorderColor /></div>
          </Stack>

        </Collapse.Panel>

        <Collapse.Panel key="3" header="Background">
          <ContainerBackgroundColor title="Background color" />
        </Collapse.Panel>

      </Collapse>
    </AttributesPanelWrapper>
  );
}
