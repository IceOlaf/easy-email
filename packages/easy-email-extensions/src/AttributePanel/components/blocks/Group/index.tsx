import React from 'react';
import { Stack } from 'easy-email-editor';
import { Width } from '@extensions/AttributePanel/components/attributes/Width';
import { BackgroundColor } from '@extensions/AttributePanel/components/attributes/BackgroundColor';
import { VerticalAlign } from '@extensions/AttributePanel/components/attributes/VerticalAlign';
import { Collapse } from 'antd';
import { AttributesPanelWrapper } from '@extensions/AttributePanel/components/attributes/AttributesPanelWrapper';

export function Group() {
  return (

    <AttributesPanelWrapper>

      <Collapse defaultActiveKey={['0', '1', '2']}>

        <Collapse.Panel key="0" header="Dimension">
          <Stack vertical spacing="tight">

            <Stack wrap={false}>
              <Stack.Item fill>
                <Width />
              </Stack.Item>
              <VerticalAlign />
            </Stack>

          </Stack>

        </Collapse.Panel>
        <Collapse.Panel key="1" header="Background">
          <Stack vertical spacing="tight">
            <BackgroundColor />
          </Stack>
        </Collapse.Panel>

      </Collapse>

    </AttributesPanelWrapper>
  );
}
