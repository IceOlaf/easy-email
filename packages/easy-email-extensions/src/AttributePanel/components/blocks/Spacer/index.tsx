import React from 'react';
import { Height } from '@extensions/AttributePanel/components/attributes/Height';
import { ContainerBackgroundColor } from '@extensions/AttributePanel/components/attributes/ContainerBackgroundColor';
import { Padding } from '@extensions/AttributePanel/components/attributes/Padding';
import { AttributesPanelWrapper } from '@extensions/AttributePanel/components/attributes/AttributesPanelWrapper';
import { Collapse } from 'antd';
import { Stack } from 'easy-email-editor';

export function Spacer() {
  return (
    <AttributesPanelWrapper>
      <Collapse defaultActiveKey={['-1', '0', '1', '2', '3']}>

        <Collapse.Panel key="1" header="Dimension">
          <Stack vertical>
            <Height />
            <Padding />
          </Stack>
        </Collapse.Panel>

        <Collapse.Panel key="2" header="Background">
          <ContainerBackgroundColor title="Background color" />
        </Collapse.Panel>

      </Collapse>
    </AttributesPanelWrapper>
  );
}
