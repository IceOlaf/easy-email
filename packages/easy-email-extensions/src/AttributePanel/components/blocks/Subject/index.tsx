import { TextField } from '@extensions';
import React from 'react';

export function Subject() {
  return (
    <TextField
      label={t('Subject')}
      name={'subject'}
      inline
    />
  )
}
