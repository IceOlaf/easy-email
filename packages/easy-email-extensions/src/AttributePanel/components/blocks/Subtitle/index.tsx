import { TextField } from '@extensions';
import React from 'react';

export function SubTitle() {
  return (
    <TextField
      label={t('SubTitle')}
      name={'subTitle'}
      inline
    />
  )
}