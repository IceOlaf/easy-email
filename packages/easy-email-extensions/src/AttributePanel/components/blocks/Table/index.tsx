import { AttributesPanelWrapper } from '@extensions/AttributePanel';
import { Button, Collapse, Tooltip } from 'antd';
import { IconFont, Stack } from 'easy-email-editor';
import React, { useState } from 'react';
import { Border } from '../../attributes/Border';
import { Color } from '../../attributes/Color';
import { ContainerBackgroundColor } from '../../attributes/ContainerBackgroundColor';
import { FontFamily } from '../../attributes/FontFamliy';
import { FontSize } from '../../attributes/FontSize';
import { FontStyle } from '../../attributes/FontStyle';
import { Padding } from '../../attributes/Padding';
import { TextAlign } from '../../attributes/TextAlign';
import { Width } from '../../attributes/Width';
import { HtmlEditor } from '../../UI/Help/HtmlEditor';

export function Table() {
  const [visible, setVisible] = useState(false);

  return (
    <AttributesPanelWrapper
      extra={
        <Tooltip title='Edit'>
          <Button
            onClick={() => setVisible(true)}
            icon={<IconFont iconName='icon-html' />}
          />
        </Tooltip>
      }
    >
      <Collapse defaultActiveKey={['-1', '0', '1', '2', '3']}>
        <Collapse.Panel key='1' header='Dimension'>
          <Stack>
            <Width />
            <Stack.Item />
          </Stack>
          <Stack vertical>
            <Padding />
          </Stack>
        </Collapse.Panel>

        <Collapse.Panel key='2' header='Decoration'>
          <Color />
          <ContainerBackgroundColor />
          <Border />
        </Collapse.Panel>

        <Collapse.Panel key='2' header='Typography'>
          <Stack>
            <FontFamily />
            <FontSize />
          </Stack>
          <FontStyle />
          <TextAlign />
        </Collapse.Panel>
      </Collapse>
      <HtmlEditor visible={visible} setVisible={setVisible} />
    </AttributesPanelWrapper>
  );
}
