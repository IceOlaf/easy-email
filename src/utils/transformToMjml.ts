import { BasicType } from '@/constants';
import { IBlockData } from '@/typings';
import { getChildIdx, getNodeIdxClassName, getNodeTypeClassName } from './block';
import { classnames } from './classnames';

export function transformToMjml(data: IBlockData, idx?: string): string {
  const att = {
    ...data.attribute,
  };
  if (idx) {
    att['css-class'] = classnames('email-block', getNodeIdxClassName(idx), getNodeTypeClassName(data.type));
  }
  const attributeStr = Object.keys(att).map(key => `${key}="${att[key]}"`).join(' ');
  const children = data.children.map((child, index) => transformToMjml(child, idx ? getChildIdx(idx, index) : undefined));
  switch (data.type) {
    case BasicType.PAGE:
      return (
        `
        <mjml>
          <mj-head>
            <mj-breakpoint width="${data.data.value.breakpoint}" />
          </mj-head>
          <mj-body ${attributeStr}>
            ${children}
          </mj-body>
        </mjml>
        `
      );
    case BasicType.WRAPPER:
      return (
        `
          <mj-wrapper ${attributeStr}>
           ${children}
          </mj-wrapper>
        `
      );
    case BasicType.SECTION:
      return (
        `
          <mj-section ${attributeStr}>
           ${children}
          </mj-section>
        `
      );
    case BasicType.COLUMN:
      return (
        `
          <mj-column ${attributeStr}>
           ${children}
          </mj-column>
        `
      );
    case BasicType.GROUP:
      return (
        `
          <mj-group ${attributeStr}>
           ${children}
          </mj-group>
        `
      );
    case BasicType.IMAGE:
      return (
        `
          <mj-image ${attributeStr}>
           ${children}
          </mj-image>
        `
      );
    case BasicType.TEXT:
      return (
        `
          <mj-text ${attributeStr}>
          ${data.data.value?.content}
          </mj-text>
        `
      );
    case BasicType.BUTTON:
      return (
        `
          <mj-button ${attributeStr}>
          ${data.data.value?.content}
          </mj-button>
        `
      );
    case BasicType.DIVIDER:
      return (
        `
          <mj-divider ${attributeStr} />
        `
      );
  }

  return '';

}