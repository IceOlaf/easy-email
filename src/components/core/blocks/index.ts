import { Page } from './basic/Page';
import { Section } from './basic/Section';
import { Column } from './basic/Column';
import { Text } from './basic/Text';
import { Image } from './basic/Image';
import { Group } from './basic/Group';
import { Button } from './basic/Button';
import { Divider } from './basic/Divider';
import { Wrapper } from './basic/Wrapper';
import { Spacer } from './basic/Spacer';

import { Raw } from './basic/Raw';
import { Accordion } from './basic/Accordion';
import { AccordionElement } from './basic/AccordionElement';
import { AccordionTitle } from './basic/AccordionTitle';
import { AccordionText } from './basic/AccordionText';
import { Carousel } from './basic/Carousel';
import { Table } from './basic/Table';
import { Hero } from './basic/Hero';
import { Navbar } from './basic/Navbar';
import { Social } from './basic/Social';
import { IBlock } from '@/typings';
import { BlockType } from '@/constants';

export class BlocksMap {
  private static blocks: IBlock[] = [
    Page,
    Section,
    Column,
    Text,
    Image,
    Group,
    Button,
    Divider,
    Wrapper,
    Spacer,
    Raw,
    Accordion,
    AccordionElement,
    AccordionTitle,
    AccordionText,
    Carousel,
    Hero,
    Navbar,
    Social,
    // TODO:
    Table,
  ];

  static registerBlock(block: IBlock) {
    this.blocks.push(block);
  }

  static findBlockByType(type: BlockType): IBlock {
    return this.blocks.find((child) => {
      return child?.type === type;
    }) as any;
  }

  static getBlocks() {
    return this.blocks;
  }
}
