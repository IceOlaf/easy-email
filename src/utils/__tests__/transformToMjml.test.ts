import { BlocksMap } from '@/components/core/blocks';
import { transformToMjml } from '../transformToMjml';

describe.only('Test transformToMjml', () => {
  it('should render as expected', () => {
    expect(
      transformToMjml(
        BlocksMap.basicBlocksMap.Page.createInstance({
          children: [
            BlocksMap.basicBlocksMap.Section.createInstance({
              children: [
                BlocksMap.basicBlocksMap.Column.createInstance({
                  children: [BlocksMap.basicBlocksMap.Text.createInstance({})],
                }),
              ],
            }),
          ],
        })
      )
    ).toMatchSnapshot();
  });
});
