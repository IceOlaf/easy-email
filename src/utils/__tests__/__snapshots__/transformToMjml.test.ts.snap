// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test transformToMjml should render as expected 1`] = `
"
        <mjml>
          <mj-head>
              <mj-breakpoint width=\\"480px\\" />
            <mj-attributes>
              
              <mj-all font-family=\\"lucida Grande,Verdana,Microsoft YaHei\\" />
              <mj-text color=\\"#000000\\" />
              
            </mj-attributes>
          </mj-head>
          <mj-body background-color=\\"#efeeea\\">
            
              <mj-section padding=\\"20px 0px 20px 0px\\" border=\\"none\\" direction=\\"ltr\\" text-align=\\"center\\" background-color=\\"#ffffff\\" background-repeat=\\"repeat\\" background-size=\\"auto\\" background-position=\\"top center\\">
               
              <mj-column padding=\\"0px 0px 0px 0px\\" border=\\"none\\" vertical-align=\\"top\\">
               
          <mj-text font-size=\\"13px\\" padding=\\"10px 25px 10px 25px\\" line-height=\\"1\\" align=\\"left\\">
           Make it easy for everyone to compose emails!
          </mj-text>
        
              </mj-column>
            
              </mj-section>
            
          </mj-body>
        </mjml>
        "
`;
